### PR TITLE
virt-config, feature-gates: Control the addition of FGs to the FG list

### PR DESCRIPTION
**Pull request by EdDev**
### What this PR does

A static FG list has been maintained, defining FGs directly into the list upon definition.

This approach has several disadvantages:
- There is no strict control on adding a new FG. E.g. an addition of two FGs with the same name is possible.
- FG have to be defined in a central place and cannot be distributed to the location that is owned by a feature.

Therefore, the current implementation is refactored to use a registration function to add FGs in a controlled manner.
Duplications is avoided by using a map instead of a list.

When `RegisterFeatureGate` is called, the inputted FG is either added to the FG list or overriding an existing FG entry.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer
The FG management is currently limited to "deprecation" feature-gates. I.e. the ones who are targeted for removal eventually.

But this mechanism can be extended and used to manage also active FGs.
Following work is going to generalize the current management to all FG states (i.e. to add Alpha & Beta states).
By doing so, we will be able to auto-generate the list of Feature-Gates and their state.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-config/deprecation/macvtap.go [hunk 1]
```diff
@@ -25,6 +25,10 @@ import (
 
 const MacvtapDiscontinueMessage = "Macvtap network binding is discontinued since v1.3. Please refer to Kubevirt user guide for alternatives."
 
+func init() {
```
Comment (ormergi) [inline pkg/virt-config/deprecation/macvtap.go:28]: The current code uses kind of global var (`featureGates` slice) and init() function that initialize it every time the `deprecation` package is imported (using `init()`).
This approach makes the code challenging to follow and has more potential for side-effects (due to how `init` function acts).
I suggest to not extend the usage of it where possible (using `init()` + `RegisterFeatureGate()` at the FR file).

**Review by ormergi — COMMENTED** [review]
Thanks for the PR

Regarding the mentioned issue about the feature-gates info being centralized, note that there is a trade off to consider, it could make the code harder to follow.
Especially if `init` function usage is extended, which by itself has some limitations and should be used with caution.

File: pkg/virt-config/deprecation/macvtap.go (hunk 1 above)
Comment (EdDev) [reply pkg/virt-config/deprecation/macvtap.go:28]: > The current code uses kind of global var (`featureGates` slice) and init() function that initialize it every time the `deprecation` package is imported (using `init()`).

There is no much change from what existed so far.
Go evaluates consts, then variables and lastly `init()`.
The practical change here is that all FGs are processed through the registration function, where the data can be normalized (e.g. setting a default message) and it is not limited to be called/defined from a single location. 

> This approach makes the code challenging to follow and has more potential for side-effects (due to how `init` function acts). I suggest to not extend the usage of it where possible (using `init()` + `RegisterFeatureGate()` at the FR file).

What exactly is challenging and which side-effects are you worried about?
The practice of using `init()` is pretty common and it looks natural to be used here.

We need the FG pool to be initialized once per application (binary) and this is exactly what happens here. As we only need a single instance of it per application, we call allow the registrations to be performed before `main()` is called.

But if you see a potential problem, please be specific and share.

**Comment by EdDev** [comment]
> Regarding the mentioned issue about the feature-gates info being centralized, note that there is a trade off to consider, it could make the code harder to follow.

The FG "manager" is a service that the project uses. Relevant features can register their FGs using this service.
Centralizing has many problems in source files:
- Collision when working on different feature-gates which are not suppose to have any dependencies.
- Ownership of such files is problematic. Which SIG owns such source files?
- Changes to such infra central source files cause a project wide recompilation as many units depend on such central source files.

While it may be nice for a developer to see all the FGs that exist in the system, I see no value he may gain from looking at it from a single file.
If the information is useful, a unit test cab document all FGs registered (not that I see the value of it either) or just generate the list of FGs from all the registered FGs.

> Especially if `init` function usage is extended, which by itself has some limitations and should be used with caution.

What caution do we need to take when using `init()`?
What risks do you see?

File: pkg/virt-config/deprecation/feature-gates.go [hunk 2]
```diff
@@ -54,30 +55,52 @@ type FeatureGate struct {
...
+// In case the FG already exists (based on its name), it overrides the
+// existing FG.
+// If the feature-gate is missing a message, a default one is set.
+func RegisterFeatureGate(fg FeatureGate) {
+	if fg.Message == "" {
+		fg.Message = fmt.Sprintf(WarningPattern, fg.Name, fg.State)
+	}
+	if existingFG := lookupFeatureGate(fg.Name); existingFG != nil {
+		*existingFG = fg
+		return
```
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/virt-config/deprecation/feature-gates.go]: Initializing it like it used to with the slice 
```
featureGates := map[string]FeatureGate={
...
}
```

[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-config/deprecation/feature-gates.go]: II suggest you focus on challenging the reasoning for the change as described in the commit and PR description.
This comment implicitly says that there is no need to do this change/PR.

File: pkg/virt-config/deprecation/macvtap.go (hunk 1 above)
Comment (ormergi) [reply pkg/virt-config/deprecation/macvtap.go:28]: The current state is no better, the fact that we use init function for initializing the featureGate list.
I think we could achieve the same with more simple solution.

With this change its now possible to cross the board and modify the featureGate function from different places through the init + registaion function, its looks hacky and somewhat of misuse of init function.

IMO its not trivial solution, using an explicit functions (and maybe complimentary) type would make it easier to follow and understand, having said that I understand it require more work and re-arranging some files.

Comment (EdDev) [reply pkg/virt-config/deprecation/macvtap.go:28]: I am not convinced it is complicated, perhaps you can show why it is complicated by comparing scenario usages for adding a new FG to the pool.

**Comment by ormergi** [comment]
Regarding ownership, this is a general issue we have today in the project and  there is on going effort to address it.

Regarding the recompilation issue, I dont see how extending the usage of init + registration function solve this issue.
Form what I understand the fact that there is such comoponent as virt-config that is used all over is the root cause for it, inst it right?
As for the feature-gate parts, it seems that the problematic file is virt-config/feature-gates.go, the functions in this file are being used all over, thus changing it or anything under virt-config trigger recompiling half of the project.

Regarding the added value of having all FG info in single file:
If its spread in multiple files/dirs/packages, using init + registration or any other solution, will make you travrese the project to realize what is the feature state in Kubevirt.
Having it in one place save this and more convenient, moreover all the code in the same place no need to figure out why we have init functions all over the place.

> What caution do we need to take when using init()?
What risks do you see?

As you already know init() functions acts different then regular functions, they are executed before main() is,
in case there are multiple init() functions in the same package, the execution flow is bottom-up, which could lead to side effects if not done properly and have more potential for bugs then regular functions that executed in main().
Anyone that reads this code has to know how init function works to understand what is done in the project.

**Comment by EdDev** [comment]
> Regarding ownership, this is a general issue we have today in the project and there is on going effort to address it.

I am addressing it. I am not sure I understand what you want to say by this comment.
 
> Regarding the recompilation issue, I dont see how extending the usage of init + registration function solve this issue. Form what I understand the fact that there is such comoponent as virt-config that is used all over is the root cause for it, inst it right? As for the feature-gate parts, it seems that the problematic file is virt-config/feature-gates.go, the functions in this file are being used all over, thus changing it or anything under virt-config trigger recompiling half of the project.

If the FG registration is distributed to the parts of the project that owns them, then this central place will only contain stable logic without the need to change it all the time. Therefore, adding or removing FGs will not require changes to it and dependent components will not need to re-compile.

> Regarding the added value of having all FG info in single file: If its spread in multiple files/dirs/packages, using init + registration or any other solution, will make you traverse the project to realize what is the feature state in Kubevirt. Having it in one place save this and more convenient, moreover all the code in the same place no need to figure out why we have init functions all over the place.

I disagree.
First, the owners of the feature and FG will probably test it, assuring it is there.
Secondly, if it is important to see all FGs in one view (why?) then one can create a unit-test to show it or generate an output file that lists all FGs and their state.

Features introduced through packages can be self contained and register their interest parts to the core project infra without the need to ask the infra to know about their existence.
Logical units depend inwards to the core project infrastructure, making sure that changes in a logical unit will not cause a project wide re-deployment/re-compilation.

> > What caution do we need to take when using init()?
> > What risks do you see?
> 
> As you already know init() functions acts different then regular functions, they are executed before main() is, in case there are multiple init() functions in the same package, the execution flow is bottom-up, which could lead to side effects if not done properly and have more potential for bugs then regular functions that executed in main(). Anyone that reads this code has to know how init function works to understand what is done in the project.

That can be said on any code snippet you will provide.
Like any other code we write,  `init()` needs to be used when appropriate.
It is used by well known libraries (e.g. [flag](https://pkg.go.dev/flag)) and is known to solve needs like the one we have here in an elegant manner.

File: pkg/virt-config/deprecation/macvtap.go
Comment (EdDev) [inline pkg/virt-config/deprecation/macvtap.go:1]: Perhaps a better move would be to a location that is owned by sig-network expressing the value of what was done here.

File: pkg/virt-config/deprecation/macvtap.go (hunk 1 above)
Comment (enp0s3) [reply pkg/virt-config/deprecation/macvtap.go:28]: @EdDev I will try to rephrase the concern - it seems like the right way to use `init()` is either in the `main` package of a program, or if the package is some kind of a service like `net/http/pprof` and then it makes sense to import it only once, from the main program source file. I am not sure though what can be harmful with redundant calls of `init()` in our case, we can use `once.Do(f)` inside the `init` to make sure the code runs only once.
@ormergi The usage of `init` was there before this PR, is this necessary to improve it in this PR?

Comment (EdDev) [reply pkg/virt-config/deprecation/macvtap.go:28]: @enp0s3 , I think it is the opposite: `init()` is mainly used in libraries, not in `main()` or `services`.
Given a library, the usage of `init()` is equivalent to defining a variable with some initial data in it. It happens once per execution and does not require an explicit call from the main flow.

A package can be imported many times, but the `init()` is still called only once, the same way a const or a variable is evaluated only once.
If two `init()` register the same FG, then the last one overrides the prev. But this is a controlled decision, we could have chosen to blow up or ignore it. This is no different from calling all this from a regular function.

**Pull request closed without merging**